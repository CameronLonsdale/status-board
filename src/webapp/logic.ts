import * as async from 'async';
import * as fs from 'fs';
import * as path from 'path';

import { readJSONFile } from '../helpers';
import { getFirst } from '../item-manager';
import logger from '../logger';

function getSafeItemName(itemName: any) {
  return path.basename(itemName).split('.')[0];
}

// ---------------------------------------------------------------
// Render custom JS for a dashboard
// ---------------------------------------------------------------
export function renderJsDashboard(packagesPath: any,
                                  wallboardAssetsFolder: any,
                                  dashboardName: any,
                                  req: any,
                                  res: any) {
  function pipeCustomJSFileNameToResponse(fileName: any, cb: any) {
    const assetFullPath = path.join(wallboardAssetsFolder, '/javascripts/', fileName);
    fs.readFile(assetFullPath, (error, fileContent) => {
      if (error) {
        logger().error(assetFullPath + ' not found');
      } else {
        res.write(fileContent + '\n\n');
      }
      cb(null);
    });
  }

  const safeDashboardName = getSafeItemName(dashboardName);

  getFirst(
    packagesPath,
    safeDashboardName,
    'dashboards',
    '.json',
    (err: any, dashboardPath: any) => {
      if (err || !dashboardPath) {
        return res.status(err ? 400 : 404).send(err ? err : 'Trying to render dashboard ' +
          safeDashboardName + ', but couldn\'t find any dashboard in the packages folder');
      }
      readJSONFile(dashboardPath, (error: any, dashboardJSON: any) => {
        if (error) {
          return res.status(400).send('Error reading dashboard');
        }
        res.type('application/javascript');
        async.eachSeries(
          (dashboardJSON.layout.customJS || []),
          pipeCustomJSFileNameToResponse,
          () => {
            res.end();
          });
      });
    });
}

export function log(req: any, res: any) {
  res.render(path.join(__dirname, '../..', 'templates', 'dashboard-log.ejs'));
}
