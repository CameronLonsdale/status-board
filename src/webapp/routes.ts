import * as path from 'path';

import configManager from '../config-manager';
import {
  log,
  renderJsDashboard,
} from './logic';
import {
  listAllDashboards,
  renderDashboard,
} from './routes/dashboard';
import {
  renderHtmlWidget,
  renderJsWidget,
  renderWidgetResource,
} from './routes/widget';

export default function (app: any, packagesPath: any) {
  const wallboardAssetsFolder = path.join(process.cwd(), 'assets');

  // -----------------------------------------
  //  Log
  // -----------------------------------------
  app.route('/log')
    .get((req: any, res: any) => {
      if (configManager('logging').liveLoggingWebAccess) {
        log(req, res);
      } else {
        // tslint:disable-next-line max-line-length
        res.status(403).end('Live logging it disabled. It must be enabled in the "logging" configuration file');
      }
    });

  // -----------------------------------------
  //  Resources for specific widget
  // -----------------------------------------
  app.route('/widgets/resources')
    .get((req: any, res: any) => {
      renderWidgetResource(path.join(process.cwd(), 'packages'), req.query.resource, req, res);
    });

  // -----------------------------------------
  //  JS for a specific widget
  // -----------------------------------------
  app.route('/widgets/:widget/js')
    .get((req: any, res: any) => {
      renderJsWidget(packagesPath, req.params.widget, req, res);
    });

  // -----------------------------------------
  //  HTML and CSS for a specific widget
  // -----------------------------------------
  app.route('/widgets/:widget')
    .get((req: any, res: any) => {
      renderHtmlWidget(packagesPath, req.params.widget, req, res);
    });

  // -----------------------------------------
  //  Dashboard
  // -----------------------------------------
  app.route('/:dashboard')
    .get((req: any, res: any) => {
      renderDashboard(packagesPath, req.params.dashboard, req, res);
    });

  // -----------------------------------------
  //  Dashboard JS
  // -----------------------------------------
  app.route('/:dashboard/js')
    .get((req: any, res: any) => {
      renderJsDashboard(packagesPath, wallboardAssetsFolder, req.params.dashboard, req, res);
    });

  // -----------------------------------------
  // List all available dashboards
  // -----------------------------------------
  app.route('/')
    .get((req: any, res: any) => {
      listAllDashboards(packagesPath, req, res);
    });
}
