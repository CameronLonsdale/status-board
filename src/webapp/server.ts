import * as bodyParser from 'body-parser';
import * as compression from 'compression';
import * as errorhandler from 'errorhandler';
import * as express from 'express';
import * as http from 'http';
import * as methodOverride from 'method-override';
import * as morgan from 'morgan';
import * as path from 'path';

import configManager from '../config-manager';
import { logger } from '../logger';
import stylus from '../stylus';
import routes from './routes';

export default function (app: express.Application, options: any) {
  http.globalAgent.maxSockets = 100;

  const atlasboardAssetsFolder = path.join(__dirname, '../../assets');
  const localAssetsFolder = path.join(process.cwd(), 'assets');

  const compiledAssetsFolder = path.join(localAssetsFolder, 'compiled');

  app.set('port', options.port);

  app.use(morgan(configManager('logging').morgan));
  app.use(compression());
  app.use(bodyParser.urlencoded({ extended: false }));
  app.use(bodyParser.json());
  app.use(methodOverride());
  app.use(errorhandler());

  app.use(stylus().getMiddleware({
    dest: compiledAssetsFolder,
    src: atlasboardAssetsFolder,
  }));

  // -----------------------------------------
  //  Expose both wallboard and Atlasboard assets.
  //  Local wallboard assets take precedence
  // -----------------------------------------
  app.use(express.static(localAssetsFolder));
  app.use(express.static(compiledAssetsFolder));
  app.use(express.static(atlasboardAssetsFolder));

  routes(app, options.packageLocations);

  const httpServer = http.createServer(app).listen(app.get('port'));
  if (!app.get('port')) {
    logger.error(`Error binding http server to port ${options.port}`);
  }
  return httpServer;
}
