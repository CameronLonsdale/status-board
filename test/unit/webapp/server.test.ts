import { server } from '../../../src/webapp/server';

describe('Webapp: Server', () => {
  test('Should match snapshot', () => {
    expect(server).toMatchSnapshot();
  });
});
