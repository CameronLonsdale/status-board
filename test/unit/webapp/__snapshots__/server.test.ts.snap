// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webapp: Server Should match snapshot 1`] = `[Function]`;
